linters/importorder/main.go

```diff
@@ -0,0 +1,465 @@
+type importDecl struct {
+	Groups []importGroup
+}
+
+// importGroup is a collection of imports
+type importGroup struct {
+	Imports []importSpec
+}
+
+// importSpec is a single import
+type importSpec struct {
+	Position token.Position
+	Line     int
+	Name     string
+	Path     string
+}
+
+// Imports returns the file imports grouped by paragraph.
+func imports(fset *token.FileSet, f *ast.File) []importDecl {
+	var importDecls []importDecl
+
+	for _, decl := range f.Decls {
+		genDecl, ok := decl.(*ast.GenDecl)
+		if !ok || genDecl.Tok != token.IMPORT {
+			continue
+		}
+
+		var (
+			importDecl importDecl
+			group      importGroup
+		)
+
+		var lastLine int
+		for _, spec := range genDecl.Specs {
+			importSpec := spec.(*ast.ImportSpec)
+			pos := importSpec.Path.ValuePos
+			line := fset.Position(pos).Line
+			fileName := fset.Position(pos).Filename
+			if lastLine > 0 && pos > 0 && line-lastLine > 1 {
+				importDecl.Groups = append(importDecl.Groups, group)
+				group = importGroup{}
+			}
+			group.Imports = append(group.Imports, newImportSpec(importSpec, line, fileName))
+			lastLine = line
+		}
+		importDecl.Groups = append(importDecl.Groups, group)
+		importDecls = append(importDecls, importDecl)
+	}
+
+	return importDecls
+}
+
+func filterOutVendor(importPaths []string) []string {
+	filteredStrings := []string{}
+	for _, importPath := range importPaths {
+		if !strings.Contains(importPath, "/vendor/") {
+			filteredStrings = append(filteredStrings, importPath)
+		}
+	}
+	return filteredStrings
+}
+
+func newImportSpec(is *ast.ImportSpec, line int, filename string) importSpec {
+	var (
+		pathLit = is.Path
+		path    string
+	)
+
+	if pathLit != nil {
+		path = pathLit.Value
+	}
+
+	return importSpec{
+		Name: filename,
+		Path: path,
+		Line: line,
+	}
+}
+
+func validateImportDecl(importDecls []importDecl) {
+	for _, importDecl := range importDecls {
+		if len(importDecl.Groups) == 0 {
+			log.Fatal(fmt.Errorf("import group cannot be empty"))
+		}
+		for _, group := range importDecl.Groups {
+			if len(group.Imports) == 0 {
+				log.Fatal(fmt.Errorf("imports cannot be empty"))
+			}
+		}
+	}
+}
+
+func findErrors(importDecls []importDecl, patterns []string, file *ast.File) lintErrors {
+	var (
+		lintErrors lintErrors
+
+		// seenPatterns is used to keep track of patterns that have been seen already
+		seenPatterns []string
+	)
+
+	validateImportDecl(importDecls)
+
+	// check if there is more than one import declaration
+	if len(importDecls) > 1 {
+		lintErr := lintError{
+			fileName: importDecls[0].Groups[0].Imports[0].Name,
+			err:      errMultipleImport,
+		}
+		lintErrors = append(lintErrors, lintErr)
+		return lintErrors
+	}
+
+	if len(importDecls) > 0 {
+		return findRecErr(importDecls[0].Groups, patterns, lintErrors, seenPatterns)
+	}
+
+	return lintErrors
+}
+
+func findRecErr(importGroup []importGroup, currentPatterns []string, lintErrors lintErrors, seenPatterns []string) lintErrors {
```

Comment:
still curious, what's up with `seenPatterns`?

Reply:
`seenPatterns` is used to keep track of the patterns that we've already checked against. For example, let's say your last import group should be `EXTERNAL`, but a previous group is `github.com/m3db/m3coordinator`, how would you know that `"github.com/m3db/m3coordinator/models"` belongs to the external group or the other one.